torchtune/models/flamingo/_encoders.py

```diff
@@ -0,0 +1,121 @@
+        num_heads: int,
+        proj_in: int,
+        proj_out: int,
+    ) -> None:
+        super().__init__()
+
+        mlp_ratio = 4
+        hidden_dim = int(mlp_ratio * embed_dim)
+        head_dim = embed_dim // num_heads
+        num_kv_heads = num_heads
+
+        transformer_layers = []
+        for idx in range(1, num_layers + 1):
+            self_attn = GroupedQueryAttention(
+                embed_dim=embed_dim,
+                num_heads=num_heads,
+                num_kv_heads=num_heads,
+                head_dim=head_dim,
+                q_proj=nn.Linear(embed_dim, num_heads * head_dim, bias=True),
+                k_proj=nn.Linear(embed_dim, num_kv_heads * head_dim, bias=True),
+                v_proj=nn.Linear(embed_dim, num_kv_heads * head_dim, bias=True),
+                output_proj=nn.Linear(embed_dim, embed_dim, bias=True),
+                pos_embeddings=None,
+                attn_dropout=0.0,
+                default_causal_mask=False,
+            )
+
+            mlp = FlamingoMLP(
+                in_dim=embed_dim,
+                hidden_dim=hidden_dim,
+                out_dim=embed_dim,
+                act_layer=nn.GELU,
+            )
+
+            layer = TransformerSelfAttentionLayer(
+                attn=attn,
+                mlp=mlp,
+                attn_norm=Fp32LayerNorm(embed_dim, eps=1e-5),
+                mlp_norm=Fp32LayerNorm(embed_dim, eps=1e-5),
+                attn_scale=TanhGate(),
+                mlp_scale=TanhGate(),
+            )
+
+            transformer_layers.append(layer)
+
+        self.projection = nn.Linear(proj_in, proj_out)
+
+    def forward(self, x: torch.Tensor, hidden_states: torch.Tensor) -> torch.Tensor:
+
+        bsz, n_ims, n_tiles, n_tokens, embed_dim = x
+
+        # transformer
+        x = x.view(bsz * n_ims, n_tiles * n_tokens, embed_dim)
+        for layer_idx, transformer_layer in enumerate(self.transformer_layers):
+            x = transformer_layer(x)
+
+        # projection
+        x = x.view(bsz, n_ims, n_tiles, n_tokens, embed_dim)
+        x = torch.cat([x, hidden_states], dim=-1)
+        x = self.projection(x)
+
+        return x
+
+
+class FlamingoVisionEncoder(nn.Module):
+    def __init__(self, vision_encoder: nn.Module, adapter: nn.Module) -> None:
+        super().__init__()
+        self.vision_encoder = vision_encoder
+        self.adapter = adapter
+
+    def forward(
+        self, images: torch.Tensor, aspect_ratio: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        x, hidden_states = self.vision_encoder(images, aspect_ratio)
+        x = self.adapter(x, hidden_states)
+        return x
+
+
+class FlamingoMLP(nn.Module):
+    def __init__(
+        self,
+        in_dim: int,
+        hidden_dim: int,
+        out_dim: int,
+        act_layer: nn.Module,
+        dropout=0.0,
+    ):
+        super().__init__()
+        self.layer1 = nn.Linear(in_dim, hidden_dim)
+        self.layer2 = nn.Linear(hidden_dim, out_dim)
```

Comment:
Any reason we can't pass in nn.Modules to the class as in the [existing FeedForward class](https://github.com/pytorch/torchtune/blob/069b12bef0b9cf735d5fb7cdc4192bfbf9abd764/torchtune/modules/feed_forward.py#L25-L27)? That'll make it easier for us to provide builders for LoRA-fied versions of these